platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ExceptionProblemRegistry.java

```diff
@@ -35,14 +40,119 @@
-
+    private final Multimap<Throwable, Problem> problemsForThrowables = Multimaps.synchronizedMultimap(MultimapBuilder.linkedHashKeys().linkedHashSetValues().<Throwable, Problem>build());
 
     public void onProblem(Throwable exception, Problem problem) {
         problemsForThrowables.put(exception, problem);
     }
 
-    public Map<Throwable, Collection<Problem>> getProblemsForThrowables() {
-        return problemsForThrowables.asMap();
+    public ProblemLookup getProblemLookup() {
+        return new DefaultProblemLookup();
+    }
+
+    /*
+     * Workaround for the fact that the exception thrown by the worker is not the same instance as the one that was thrown by the build. With the lookup we can find the original exception by comparing
+     * the stack frames. The comparison is expensive, so we only do when it's necessary (when the original exception does not contain the target and there's a matching class name and message).
+     */
+    private class DefaultProblemLookup implements ProblemLookup {
+
+
+        private final Multimap<String, Throwable> lookup;
+
+        DefaultProblemLookup() {
+            this.lookup = initLookup(problemsForThrowables.keySet());
+        }
+
+        private Multimap<String, Throwable> initLookup(Set<Throwable> exceptions) {
+            Multimap<String, Throwable> lookup = ArrayListMultimap.create();
+            for (Throwable exception : exceptions) {
+                lookup.put(key(exception), exception);
+            }
+            return lookup;
+        }
+
+        private String key(Throwable t) {
+            return t.getClass().getName() + ":" + messageOf(t);
+        }
+
+        private String messageOf(Throwable t) {
+            String result = "";
+            try {
+                String message = t.getMessage();
+                result = message == null ? "" : message;
+            } catch (RuntimeException ignore) {
+                // ignore exceptions with faulty getMessage() implementation
+            }
+            return result;
+        }
+
+        @Override
+        public Collection<Problem> findAll(Throwable t) {
+            Throwable throwable = find(t);
+            return throwable == null ? ImmutableList.<Problem>of() : ImmutableList.copyOf(problemsForThrowables.get(throwable));
+        }
+
+        @Nullable
+        private Throwable find(Throwable t) {
+            try {
+                if (problemsForThrowables.keySet().contains(t)) {
+                    return t;
+                }
+                Collection<Throwable> candidates = lookup.get(key(t));
+                for (Throwable candidate : candidates) {
+                    if (deepEquals(candidate, t, new ArrayList<Throwable>())) {
+                        return candidate;
+                    }
+                }
+            } catch (RuntimeException e) {
+                e.printStackTrace();
```

Comment:
❌ I don't think we should use `e.printStackTrace()` in our code base. Not sure if you want to do some debug logging here or what is the actual problem we guard against. I'd just throw the runtime exception if we don't have anything in particular to guard against here.

Reply:
I agree, the `printStackTrace()` call should be deleted. I'd rather not rethrow the exception. We compare exception instances with user-supplied types that can have exotic problems that we don't anticipate. If the comparison fails, I'd not break the control flow but continue with returning null.

Reply:
How can we find out that the exotic problems (or any problems) happen? Should we print out the problem on the info log?